libyul/optimiser/ValueConstraintBasedSimplifier.cpp

```diff
@@ -0,0 +1,384 @@
+		case dev::solidity::Instruction::CREATE:
+		case dev::solidity::Instruction::CREATE2:
+			return ValueConstraint::bitRange(159, 0);
+
+		case dev::solidity::Instruction::CALL:
+		case dev::solidity::Instruction::CALLCODE:
+		case dev::solidity::Instruction::DELEGATECALL:
+		case dev::solidity::Instruction::STATICCALL:
+			return ValueConstraint::boolean();
+
+		case dev::solidity::Instruction::KECCAK256:
+		case dev::solidity::Instruction::EXTCODEHASH:
+		case dev::solidity::Instruction::BLOCKHASH:
+		case dev::solidity::Instruction::MLOAD:
+		case dev::solidity::Instruction::SLOAD:
+
+		// These could be restricted in a real-world setting:
+		case dev::solidity::Instruction::BALANCE:
+		case dev::solidity::Instruction::CALLVALUE:
+		case dev::solidity::Instruction::CALLDATASIZE:
+		case dev::solidity::Instruction::GASPRICE:
+		case dev::solidity::Instruction::EXTCODESIZE:
+		case dev::solidity::Instruction::RETURNDATASIZE:
+		case dev::solidity::Instruction::TIMESTAMP:
+		case dev::solidity::Instruction::NUMBER:
+		case dev::solidity::Instruction::DIFFICULTY:
+		case dev::solidity::Instruction::GASLIMIT:
+		case dev::solidity::Instruction::PC:
+		case dev::solidity::Instruction::MSIZE:
+		case dev::solidity::Instruction::GAS:
+		default:
+			return {};
+		}
+	}
+
+private:
+
+	map<YulString, ValueConstraint> const& m_variableConstraints;
+};
+}
+
+ValueConstraint ValueConstraint::constant(u256 const& _value)
+{
+	return ValueConstraint{_value, _value, _value, _value};
+}
+
+ValueConstraint ValueConstraint::boolean()
+{
+	return ValueConstraint{0, 1, 0, 1};
+}
+
+ValueConstraint ValueConstraint::bitRange(size_t _highest, size_t _lowest)
+{
+	yulAssert(_highest >= _lowest, "");
+	return valueFromBits(
+		0,
+		(u256(-1) >> (255 - _highest)) & ~(u256(-1) >> (256 - _lowest))
+	);
+}
+
+ValueConstraint ValueConstraint::valueFromBits(u256 _minBits, u256 _maxBits)
+{
+	int highestBit = highestBitSet(_maxBits);
+	return ValueConstraint{
+		0,
+		u256(-1) >> (255 - highestBit),
+		move(_minBits),
+		move(_maxBits)
+	};
+}
+
+ValueConstraint ValueConstraint::bitsFromValue(u256 _minValue, u256 _maxValue)
+{
+	int highestBit = highestBitSet(_maxValue);
+	return ValueConstraint{
+		move(_minValue),
+		move(_maxValue),
+		0,
+		u256(-1) >> (255 - highestBit)
+	};
+}
+
+ValueConstraint ValueConstraint::operator&(ValueConstraint const& _other)
+{
+	return valueFromBits(
+		minBits & _other.minBits,
+		maxBits & _other.maxBits
+	);
+}
+
+ValueConstraint ValueConstraint::operator|(ValueConstraint const& _other)
+{
+	return valueFromBits(
+		minBits | _other.minBits,
+		maxBits | _other.maxBits
+	);
+}
+
+ValueConstraint ValueConstraint::operator~()
+{
+	return valueFromBits(~maxBits, ~minBits);
+}
+
+ValueConstraint ValueConstraint::operator+(ValueConstraint const& _other)
+{
+	if (bigint(maxValue) + bigint(_other.maxValue) > u256(-1))
+		return ValueConstraint{}; // overflow
+	else
+		return bitsFromValue(
+			minValue + _other.minValue,
```

Comment:
Can this also overflow?

Reply:
Ah, if this one overflows the other one also overflows, and it already returns, so it's fine.